<a href="https://colab.research.google.com/github/Nicolaser1102/Introduccion_to_SQL/blob/main/Ejercicios_SQL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CODING DOJO 
## Fecha creación:31/1/2022
## Fecha de actualización: 31/1/2022
## Esteban Simbaña

#Ejercicios SQL

## *Plantilla para montar el entorno de Python, tomada de Coding Dojo*

In [1]:
# Importamos las librerías que necesitaremos
import pandas as pd
from sqlalchemy import create_engine

In [3]:
# Montamos los datos
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [34]:
# Cargamos tabla 1 (school_details)
school = pd.read_csv('/content/drive/MyDrive/Coding Dojo/Códigos/Track 3/Deberes /Week 4 Track 3/School_Details.csv')
school.head(2)

DistrictName EstablishmentCode  \
0         East             S0169   
1         East             S0099   

                                   EstablishmentName   BRCName  \
0               BUDANG GOVERNMENT PRIMARY SCHOOL (E)  Duga Brc   
1  CENTRAL PENDAM GOVERNMENT SENIOR SECONDARY SCHOOL  Duga Brc   

                                             CRCName SchoolLevel  \
0  Central Pendam Government Senior Secondary School          PS   
1  Central Pendam Government Senior Secondary School         SSS   

  ManagementName LocalityHabitation SubdivisionName ConstituencyName  ...  \
0    State Govt.             BUDANG         Gangtok          Rhenock  ...   
1    State Govt.     CENTRAL PENDAM         Gangtok      West Pandam  ...   

  EstablishedDate DistrictHQDistance (in Km) CRCDistance (in Km)  \
0           28126                       58.0                 8.0   
1           15342                       50.0                 0.0   

  BACDistance (in Km) IsRecognized RecognizedBy ContactPersonName LowestClass  \
0                16.0   Recognized         HRDD               NaN     Class 1   
1                 6.0   Recognized         HRDD               NaN     Class 1   

  HighestClass BoardName  
0      Class 5       NaN  
1     Class 12  C.B.S.E.  

[2 rows x 30 columns]

In [35]:
# Cargamos tabla 2 (school_facility_details)
facility = pd.read_csv('/content/drive/MyDrive/Coding Dojo/Códigos/Track 3/Deberes /Week 4 Track 3/School_Facility_Details.csv')
facility.head(2)

EstablishmentCode                                  EstablishmentName  \
0             S1034  (I.E.C) INDIAN EVANGALISTIC CRUSADE MISSION PR...   
1             S0516                10TH MILE GOVERNMENT PRIMARY SCHOOL   

  IsLibraryAvailable  AvailableBooksCount IsFireExtgAvailable  \
0      Not Available                    0           Available   
1      Not Available                    0       Not Available   

   ExtinguisherCount HOIRoomAvailable IsStaffRoomAvailable  NoofStaffRoom  \
0                  1        Available        Not Available              0   
1                  0        Available            Available              1   

  IsLockerForStaffAvailable IsSportsStoreRoomAvailable  
0             Not Available              Not Available  
1             Not Available              Not Available

In [7]:
# Creamos sqlite engine
engine2 = create_engine('sqlite:///school.db', echo=True)
# Creamos una conexión a engine

conn = engine2.connect()

In [8]:
# Añadimos la base de datos como una tabla en sqlite
sqlite_table = "school_details"
school.to_sql(sqlite_table, conn, if_exists='fail')

sqlite_table2 = "school_facility_details"
facility.to_sql(sqlite_table2, conn, if_exists='fail')

2022-03-31 23:52:14,949 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("school_details")
2022-03-31 23:52:14,951 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-03-31 23:52:14,953 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("school_details")
2022-03-31 23:52:14,955 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-03-31 23:52:14,970 INFO sqlalchemy.engine.Engine 
CREATE TABLE school_details (
	"index" BIGINT, 
	"DistrictName" TEXT, 
	"EstablishmentCode" TEXT, 
	"EstablishmentName" TEXT, 
	"BRCName" TEXT, 
	"CRCName" TEXT, 
	"SchoolLevel" TEXT, 
	"ManagementName" TEXT, 
	"LocalityHabitation" TEXT, 
	"SubdivisionName" TEXT, 
	"ConstituencyName" TEXT, 
	"RevenueBlockName" TEXT, 
	"CivicBodyName" TEXT, 
	"WardName" TEXT, 
	"PostOfficeName" TEXT, 
	"PhoneNo" TEXT, 
	"FaxNo" TEXT, 
	"EmailID" TEXT, 
	"Website" TEXT, 
	"LocatedInRuralOrUrban" TEXT, 
	"DiseCode" TEXT, 
	"EstablishedDate" TEXT, 
	"DistrictHQDistance (in Km)" FLOAT, 
	"CRCDistance (in Km)" FLOAT, 
	"BACDistance (in 

## **Ahora, escribe consultas SQL para responder las siguientespreguntas:**

## Aggregates

### 1. ¿Cuál es el BACDistance promedio (en km)? Renombra la columna resultante como "avg_bac_distancia."

In [25]:
#verificamos cuantas filas tiene la dataframe original
school.shape

(1249, 30)

In [28]:
#obtenemos el promedio manualmente de la dataframe en pandas
(school['BACDistance (in Km)']).sum()/1249

12.651361088871097

In [15]:
# Write your SQL query in this string
sql = """
select AVG("BACDistance (in Km)") as avg_bac_distance

from school_details"""

# Read the sql query into pandas to view the results
df = pd.read_sql_query(sql, engine2)
df.head()

2022-03-31 23:55:10,848 INFO sqlalchemy.engine.Engine 
select AVG("BACDistance (in Km)") as avg_bac_distance

from school_details
2022-03-31 23:55:10,851 INFO sqlalchemy.engine.Engine [raw sql] ()


avg_bac_distance
0         12.681822

Se verifica que nuestra primera consulta sql está bien hecha comparando el resultado del promedio obtenido manualmente **12.65**, con el obtenido mediante la consulta sql **12.68**

## GROUP BY

### 2. ¿Cuál es el BACDistance promedio (en km) según nivel educativo del colegio?  Ordena los resultados de promedio mayor a promedio menor.

In [18]:
#observamos cuantos niveles de educación existen en la data frame original
school["SchoolLevel"].unique()

array(['PS', 'SSS', 'SS', 'JHS', 'LPS', nan], dtype=object)

In [21]:
# Write your SQL query in this string
sql = """

SELECT "SchoolLevel", AVG("BACDistance (in Km)") AS avg_BACDistanceProm
FROM school_details
GROUP BY "SchoolLevel"
ORDER BY "avg_BACDistanceProm" DESC
"""

# Read the sql query into pandas to view the results
df = pd.read_sql_query(sql, engine2)
df

2022-04-01 00:03:14,186 INFO sqlalchemy.engine.Engine 

SELECT "SchoolLevel", AVG("BACDistance (in Km)") AS avg_BACDistanceProm
FROM school_details
GROUP BY "SchoolLevel"
ORDER BY "avg_BACDistanceProm" DESC

2022-04-01 00:03:14,192 INFO sqlalchemy.engine.Engine [raw sql] ()


SchoolLevel  avg_BACDistanceProm
0         JHS            13.692647
1          PS            12.830787
2          SS            12.630872
3         SSS            10.863095
4         LPS             5.368421
5        None                  NaN

## Having

### 3. Repite la consulta anterior, pero solo muestra los resultados donde la distancia promedio sea menor a 10 km.

In [29]:
# Write your SQL query in this string
sql = """

SELECT "SchoolLevel", AVG("BACDistance (in Km)") AS avg_BACDistanceProm
FROM school_details
GROUP BY "SchoolLevel"

HAVING avg_BACDistanceProm < 10

ORDER BY "avg_BACDistanceProm" DESC



"""

# Read the sql query into pandas to view the results
df = pd.read_sql_query(sql, engine2)
df.head()

2022-04-01 00:08:44,817 INFO sqlalchemy.engine.Engine 

SELECT "SchoolLevel", AVG("BACDistance (in Km)") AS avg_BACDistanceProm
FROM school_details
GROUP BY "SchoolLevel"

HAVING avg_BACDistanceProm < 10

ORDER BY "avg_BACDistanceProm" DESC




2022-04-01 00:08:44,822 INFO sqlalchemy.engine.Engine [raw sql] ()


SchoolLevel  avg_BACDistanceProm
0         LPS             5.368421

## Joins

### 4. Une las dos tablas (detalles de colegios y detalles de instalaciones de los colegios). Comienza mostrando todas las columnas, pero limita las filas resultantes a solo 2.

In [60]:
# Write your SQL query in this string
sql = """

SELECT *
FROM
   school_details
INNER JOIN
   school_facility_details
ON school_details.EstablishmentCode = school_facility_details.EstablishmentCode

"""

# Read the sql query into pandas to view the results
df = pd.read_sql_query(sql, engine2)
df.head(2)

2022-04-01 00:51:07,754 INFO sqlalchemy.engine.Engine 

SELECT *
FROM
   school_details
INNER JOIN
   school_facility_details
ON school_details.EstablishmentCode = school_facility_details.EstablishmentCode


2022-04-01 00:51:07,763 INFO sqlalchemy.engine.Engine [raw sql] ()


index DistrictName EstablishmentCode  \
0      0         East             S0169   
1      1         East             S0099   

                                   EstablishmentName   BRCName  \
0               BUDANG GOVERNMENT PRIMARY SCHOOL (E)  Duga Brc   
1  CENTRAL PENDAM GOVERNMENT SENIOR SECONDARY SCHOOL  Duga Brc   

                                             CRCName SchoolLevel  \
0  Central Pendam Government Senior Secondary School          PS   
1  Central Pendam Government Senior Secondary School         SSS   

  ManagementName LocalityHabitation SubdivisionName  ...  \
0    State Govt.             BUDANG         Gangtok  ...   
1    State Govt.     CENTRAL PENDAM         Gangtok  ...   

                                   EstablishmentName IsLibraryAvailable  \
0               BUDANG GOVERNMENT PRIMARY SCHOOL (E)          Available   
1  CENTRAL PENDAM GOVERNMENT SENIOR SECONDARY SCHOOL          Available   

  AvailableBooksCount IsFireExtgAvailable ExtinguisherCount HOIRoomAvailable  \
0                 100           Available                 2        Available   
1                   0           Available                 5        Available   

  IsStaffRoomAvailable NoofStaffRoom IsLockerForStaffAvailable  \
0            Available             1             Not Available   
1            Available             2             Not Available   

  IsSportsStoreRoomAvailable  
0              Not Available  
1                  Available  

[2 rows x 43 columns]

Plantilla para realizar uniones en SQL inspirada de:
https://desarrolloweb.com/articulos/2355.php

### 5. Ahora, solo selecciona colegios donde el nivel de este sea Escuela Primaria y que no tengan biblioteca.

Escuela Primaria = "PS"

In [62]:
facility["IsLibraryAvailable"].unique()

array(['Not Available', 'Available'], dtype=object)

In [68]:
# Write your SQL query in this string
sql = """

SELECT *
FROM
   school_details
INNER JOIN
   school_facility_details
ON school_details.EstablishmentCode = school_facility_details.EstablishmentCode

WHERE SchoolLevel = "PS" AND IsLibraryAvailable = "Available"

ORDER BY ciudad DESC
 
"""

# Read the sql query into pandas to view the results
df = pd.read_sql_query(sql, engine2)
df.head(2)


2022-04-01 00:54:37,750 INFO sqlalchemy.engine.Engine 

SELECT *
FROM
   school_details
INNER JOIN
   school_facility_details
ON school_details.EstablishmentCode = school_facility_details.EstablishmentCode

WHERE SchoolLevel = "PS" AND IsLibraryAvailable = "Available"
 

2022-04-01 00:54:37,754 INFO sqlalchemy.engine.Engine [raw sql] ()


index DistrictName EstablishmentCode                     EstablishmentName  \
0      0         East             S0169  BUDANG GOVERNMENT PRIMARY SCHOOL (E)   
1      6         East             S0870   ETERNAL ENGLISH SCHOOL, WEST PENDAM   

    BRCName                                            CRCName SchoolLevel  \
0  Duga Brc  Central Pendam Government Senior Secondary School          PS   
1  Duga Brc  Central Pendam Government Senior Secondary School          PS   

      ManagementName LocalityHabitation SubdivisionName  ...  \
0        State Govt.             BUDANG         Gangtok  ...   
1  Private (Unaided)  SAKHU WEST PENDAM         Gangtok  ...   

                      EstablishmentName IsLibraryAvailable  \
0  BUDANG GOVERNMENT PRIMARY SCHOOL (E)          Available   
1   ETERNAL ENGLISH SCHOOL, WEST PENDAM          Available   

  AvailableBooksCount IsFireExtgAvailable ExtinguisherCount HOIRoomAvailable  \
0                 100           Available                 2        Available   
1                   0           Available                 1    Not Available   

  IsStaffRoomAvailable NoofStaffRoom IsLockerForStaffAvailable  \
0            Available             1             Not Available   
1        Not Available             0             Not Available   

  IsSportsStoreRoomAvailable  
0              Not Available  
1              Not Available  

[2 rows x 43 columns]

### 6. Toma la consulta anterior, pero ordena los resultados en orden alfabético por nombre.

In [70]:
# Write your SQL query in this string
sql = """

SELECT *
FROM
   school_details
INNER JOIN
   school_facility_details
ON school_details.EstablishmentCode = school_facility_details.EstablishmentCode

WHERE SchoolLevel = "PS" AND IsLibraryAvailable = "Available"

ORDER BY EstablishmentName 

"""

# Read the sql query into pandas to view the results
df = pd.read_sql_query(sql, engine2)
df.head()

2022-04-01 00:58:27,845 INFO sqlalchemy.engine.Engine 

SELECT *
FROM
   school_details
INNER JOIN
   school_facility_details
ON school_details.EstablishmentCode = school_facility_details.EstablishmentCode

WHERE SchoolLevel = "PS" AND IsLibraryAvailable = "Available"

ORDER BY EstablishmentName 


2022-04-01 00:58:27,849 INFO sqlalchemy.engine.Engine [raw sql] ()


index DistrictName EstablishmentCode  \
0    345         East             S0178   
1    385         East             S0146   
2    143         East             S0187   
3    457        North             S0067   
4   1149         West             S1217   

                            EstablishmentName       BRCName  \
0          AGRIGAON GOVERNMENT PRIMARY SCHOOL      Regu Brc   
1          ANKUCHEN GOVERNMENT PRIMARY SCHOOL   Pakyong Brc   
2  BADONG THANGSING GOVERNMENT PRIMARY SCHOOL  Khamdong Brc   
3            BAKCHA GOVERNMENT PRIMARY SCHOOL      Kabi Brc   
4               BAL JYOTI ACADEMY, TIMBURBONG    Soreng Brc   

                                             CRCName SchoolLevel  \
0              Dalapchen Government Secondary School          PS   
1                 Linkey Government Secondary School          PS   
2        Khamdong Government Senior Secondary School          PS   
3  Kalzang Gyatso Government senior Secondary School          PS   
4          Soreng Government Senior Secondary School          PS   

      ManagementName LocalityHabitation SubdivisionName  ...  \
0        State Govt.     AGRIGAON BUSTY          Rongli  ...   
1        State Govt.     ANKUCHEN BUSTY         Pakyong  ...   
2        State Govt.   BADONG THANGSING         Gangtok  ...   
3        State Govt.        SAMDONGCHEN          Mangan  ...   
4  Private (Unaided)         TIMBURBONG          Soreng  ...   

                            EstablishmentName IsLibraryAvailable  \
0          AGRIGAON GOVERNMENT PRIMARY SCHOOL          Available   
1          ANKUCHEN GOVERNMENT PRIMARY SCHOOL          Available   
2  BADONG THANGSING GOVERNMENT PRIMARY SCHOOL          Available   
3            BAKCHA GOVERNMENT PRIMARY SCHOOL          Available   
4               BAL JYOTI ACADEMY, TIMBURBONG          Available   

  AvailableBooksCount IsFireExtgAvailable ExtinguisherCount HOIRoomAvailable  \
0                  60           Available                 2    Not Available   
1                 300           Available                 2    Not Available   
2                  13           Available                 2    Not Available   
3                   6           Available                 2    Not Available   
4                   0           Available                 0    Not Available   

  IsStaffRoomAvailable NoofStaffRoom IsLockerForStaffAvailable  \
0            Available             1             Not Available   
1            Available             1             Not Available   
2            Available             1             Not Available   
3            Available             1             Not Available   
4            Available             1                 Available   

  IsSportsStoreRoomAvailable  
0              Not Available  
1              Not Available  
2              Not Available  
3              Not Available  
4                  Available  

[5 rows x 43 columns]

In [ ]:
# Close the database connection
conn.close()